### PR TITLE
Add game testing.

### PR DESCRIPTION
**Pull request by o01eg**
This PR adds tests for the game itself.
Test acts as a host for the game server. It starts singleplayer game, make a turns and check if server and AIs acts correctly and within time limits.
Test accepts environment variables `FO_TEST_GAME_AIS` to manage number of AI players, `FO_TEST_GAME_TURNS` to manage numbers of turns to make, and `FO_TEST_GAME_SAVE` to save game after each turn.
Test will stop when test player will be eliminated or someone wins the game.

Other changes:
- Added `Process::Terminate` method to ensure server was stopped correctly.
- `ClientApp::VerifyCheckSum` returns `bool` instead `void`.
- Add server flag `--testing` to output AI logs to `stdout`.

### CODE REVIEW

File: test/game/TestGame.cpp [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+                return false;
+            auto opt_msg = m_networking->GetMessage();
+            if (opt_msg) {
+                Message msg = *opt_msg;
+                if (! HandleMessage(msg)) {
+                    return false;
+                }
+                to_process = true;
+            } else {
+                if(to_process) {
```
Comment (geoffthemedio) [inline test/game/TestGame.cpp]: Space between `(` and `if`

**Comment by o01eg** [comment]
Rebased. Refactored and set time limits on game starting and turn processing.

File: test/game/TestGame.cpp (hunk 1 above)
Comment (o01eg) [reply test/game/TestGame.cpp]: Fixed.

**Comment by o01eg** [comment]
Rebased. Added environment variables `TEST_GAME_AIS` to manage number of AI players and `TEST_GAME_TURNS` to manage numbers of turns to make.

**Comment by o01eg** [comment]
Rebased. Implement testing saves. Add options `TEST_GAME_SAVE` to manage if saves will be applied (default `true`).

**Comment by o01eg** [comment]
Rebased. Tested against new docker image.

File: default/stringtables/en.txt [hunk 2]
```diff
@@ -1186,6 +1186,9 @@ Start server in hostless mode. Server accepts players in the multiplayer lobby,
 OPTIONS_DB_SKIP_CHECKSUM
 Skip comparing the resource directory checksums.  This allows faster startup when client and server are on the same machine, using the same resource directory.
 
+OPTIONS_DB_TESTING
+Start server in testing mode. AI playing will be output log to colnsole instead file (Linux only).
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: colnsole
rewrite second sentence

Not clear if this option is Linux-only, or the to-consolve vs. to-file part is Linux-only

File: server/ServerApp.cpp [hunk 3]
```diff
@@ -260,6 +260,14 @@ void ServerApp::CreateAIClients(const std::vector<PlayerSetupData>& player_setup
         args.push_back(GetOptionsDB().Get<std::string>("log-level"));
     }
 
+    if (GetOptionsDB().Get<bool>("testing")) {
+#ifdef FREEORION_LINUX
+        // Dirty hack to output log to console.
```
Comment (geoffthemedio) [inline server/ServerApp.cpp]: why not use the standard log files?

**Comment by geoffthemedio** [comment]
Can this be tested / used locally? Running `freeoriond --testing` just starts the server, which parses content and then sits doing nothing.

File: server/ServerApp.cpp (hunk 3 above)
Comment (o01eg) [reply server/ServerApp.cpp]: I'm not sure if it possible to get them from Travis CI to get detailed information about possible errors.

**Comment by o01eg** [comment]
@geoffthemedio Yes, it's a simple unittest. It could be called with `ctest` or `cmake --build . --target unittest`

File: test/game/ClientAppFixture.cpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    //InitLoggingOptionsDBSystem();
+
+    InfoLogger() << FreeOrionVersionString();
+    DebugLogger() << "Test client initialized";
+
+    StartBackgroundParsing();
+}
+
+int ClientAppFixture::EffectsProcessingThreads() const
+{ return 1; }
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/ClientAppFixture.cpp]: It requires unittest to have a fully initialized `OptionsDB`.

File: test/game/ClientAppFixture.cpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+#include <boost/format.hpp>
+
+ClientAppFixture::ClientAppFixture()
+    :m_game_started(false)
+{
+#ifdef FREEORION_LINUX
+    // Dirty hack to output log to console.
+    InitLoggingSystem("/proc/self/fd/1", "Test");
+#else
+    InitLoggingSystem((GetUserDataDir() / "test.log").string(), "Test");
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/ClientAppFixture.cpp]: `GetUserDataDir()` is a standard folder for logs. `test.log` is used instead `freeorion.log` to distinguish unit test from human client.

**Comment by geoffthemedio** [comment]
But not without CMake?

File: server/ServerApp.cpp (hunk 3 above)
Comment (geoffthemedio) [reply server/ServerApp.cpp]: Running the tests isn't possible on AppVeyor as well?

File: test/game/main.cpp [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+#include <boost/test/unit_test_monitor.hpp>
+#include <boost/test/output/compiler_log_formatter.hpp>
+
+// Produces an GCC-ish message prefix for various frontends that understand gcc output.
+struct gcc_log_formatter :
+    public boost::unit_test::output::compiler_log_formatter
+{
+    void print_prefix(std::ostream& output, boost::unit_test::const_string file_name, std::size_t line) override
+    {
+        output << file_name << ':' << line << ": error: ";
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/main.cpp]: Fixed.

File: test/game/ClientAppFixture.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+class ClientAppFixture : public ClientApp {
+public:
+    ClientAppFixture();
+
+    bool PingLocalHostServer();
+
+    bool ConnectToLocalHostServer();
+
+    void HostSPGame(unsigned int num_AIs);
+
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/ClientAppFixture.h]: Fixed.

File: test/game/ClientAppFixture.cpp [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+{ return m_networking->ConnectToLocalHostServer(); }
+
+void ClientAppFixture::HostSPGame(unsigned int num_AIs) {
+    std::vector<std::pair<std::string, std::string>> game_rules = GetGameRules().GetRulesAsStrings();
+
+    SinglePlayerSetupData setup_data;
+    setup_data.m_new_game = true;
+    setup_data.m_filename.clear();  // not used for new game
+
+    // get values stored in options from previous time game was run or
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/ClientAppFixture.cpp]: Fixed.

File: test/game/ClientAppFixture.cpp [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+{ return 1; }
+
+bool ClientAppFixture::PingLocalHostServer()
+{ return m_networking->PingLocalHostServer(std::chrono::milliseconds(100)); }
+
+bool ClientAppFixture::ConnectToLocalHostServer()
+{ return m_networking->ConnectToLocalHostServer(); }
+
+void ClientAppFixture::HostSPGame(unsigned int num_AIs) {
+    std::vector<std::pair<std::string, std::string>> game_rules = GetGameRules().GetRulesAsStrings();
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/ClientAppFixture.cpp]: Fixed.

File: test/game/ClientAppFixture.cpp [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+#include "util/Directories.h"
+#include "util/GameRules.h"
+#include "util/Version.h"
+
+#include "GG/GG/ClrConstants.h"
+
+#include <boost/format.hpp>
+
+ClientAppFixture::ClientAppFixture()
+    :m_game_started(false)
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/ClientAppFixture.cpp]: Fixed.

File: default/stringtables/en.txt (hunk 2 above)
Comment (o01eg) [reply default/stringtables/en.txt]: Option is no-op in other than Linux OS.

Comment (geoffthemedio) [reply default/stringtables/en.txt]: Put the "Linux only" at the start.

And perhaps don't even add it as an option in those cases...

**Comment by o01eg** [comment]
@geoffthemedio Without cmake it could be called with `./fo_unittest_game "--log_level=message" "--run_test=TestGame" "--catch_system_error=no"`.

**Comment by geoffthemedio** [comment]
Except there is presumably no `fo_unittest_game` to run without having built using CMake?

**Comment by o01eg** [comment]
@geoffthemedio Could other unit tests (parser and GG) be built without cmake?

File: default/stringtables/en.txt (hunk 2 above)
Comment (o01eg) [reply default/stringtables/en.txt]: Moved the option to macro condition for `FREEORION_LINUX`.

File: server/ServerApp.cpp (hunk 3 above)
Comment (o01eg) [reply server/ServerApp.cpp]: It possible but logs will be output to files.

**Comment by geoffthemedio** [comment]
I don't know

File: test/game/TestGame.cpp [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+#include "ClientAppFixture.h"
+#include "Empire/Empire.h"
+#include "universe/Enums.h"
+#include "util/Directories.h"
+#include "util/Process.h"
+#include "util/SitRepEntry.h"
+
+namespace {
+    std::string ServerClientExe() {
```
Comment (adrianbroher) [inline test/game/TestGame.cpp]: `ServerClient`

excuse me, what?

**Comment by adrianbroher** [comment]
@o01eg 
> Could other unit tests (parser and GG) be built without cmake?

No, but that doesn't matter as it's a Windows only problem.

File: test/game/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,74 @@
+find_package(Boost ${MINIMUM_BOOST_VERSION}
+    COMPONENTS
+        unit_test_framework
+        log
+        log_setup
+    REQUIRED)
+
+add_executable(fo_unittest_game
```
Comment (adrianbroher) [inline test/game/CMakeLists.txt]: This is not a unit test, but rather a smoke and system test.

Can you please rename the whole `test/game` directory to `test/system` and the executable to `fo_systemtest_game`?

File: test/game/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        ${PROJECT_SOURCE_DIR}/network/ClientNetworking.h
+    PRIVATE
+        ${PROJECT_SOURCE_DIR}/client/ClientApp.cpp
+        ${PROJECT_SOURCE_DIR}/client/ClientFSMEvents.cpp
+        ${PROJECT_SOURCE_DIR}/GG/src/ClrConstants.cpp
+        ${PROJECT_SOURCE_DIR}/network/ClientNetworking.cpp
+)
+
+set(FO_TEST_GAME
+    TestGame
```
Comment (adrianbroher) [inline test/game/CMakeLists.txt]: Can you rename this to `SmokeTestGame`?

File: test/game/TestGame.cpp (hunk 11 above)
Comment (o01eg) [reply test/game/TestGame.cpp]: I've got it from `client/human/HumanClientApp.cpp:413`

File: test/game/TestGame.cpp [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+            num_AIs = boost::lexical_cast<unsigned int>(env_num_AIs);
+        } catch (...) {
+            // ignore
+        }
+    }
+
+    const char *env_num_turns = std::getenv("TEST_GAME_TURNS");
+    if (env_num_turns) {
+        try {
+            num_turns = boost::lexical_cast<int>(env_num_turns);
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/TestGame.cpp]: `m_current_turn` is `int` so I make other turn variable `int` too. Test simply won't do any turns as 0 are greater then negative number.

File: test/game/TestGame.cpp [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+    BOOST_TEST_MESSAGE("First messages processed. Starting game...");
+
+    unsigned int num_AIs = 2;
+    int num_turns = 2;
+    bool save_game = true;
+
+    const char *env_num_AIs = std::getenv("TEST_GAME_AIS");
+    if (env_num_AIs) {
+        try {
+            num_AIs = boost::lexical_cast<unsigned int>(env_num_AIs);
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/TestGame.cpp]: I'm not sure. Is something except RAM size limits AI client number?

File: test/game/TestGame.cpp [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    constexpr static int MAX_WAITING_SEC = 120;
+}
+
+#ifdef FREEORION_MACOSX
+#include <stdlib.h>
+#endif
+
+BOOST_FIXTURE_TEST_SUITE(TestGame, ClientAppFixture)
+
+BOOST_AUTO_TEST_CASE(host_server) {
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/TestGame.cpp]: Added description to commentaries before the test code.

File: test/game/TestGame.cpp [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+    boost::posix_time::ptime start_time = boost::posix_time::microsec_clock::local_time();
+    BOOST_REQUIRE(ProcessMessages(start_time, MAX_WAITING_SEC));
+
+    BOOST_TEST_MESSAGE("First messages processed. Starting game...");
+
+    unsigned int num_AIs = 2;
+    int num_turns = 2;
+    bool save_game = true;
+
+    const char *env_num_AIs = std::getenv("TEST_GAME_AIS");
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/game/TestGame.cpp]: Added them to the test's description.

File: test/game/CMakeLists.txt (hunk 12 above)
Comment (o01eg) [reply test/game/CMakeLists.txt]: Renamed.

File: test/game/CMakeLists.txt (hunk 13 above)
Comment (o01eg) [reply test/game/CMakeLists.txt]: Renamed.

**Comment by geoffthemedio** [comment]
@adrianbroher you happy / ok with this now?

**Comment by genestack-solomatin** [comment]
I am not happy. 

I don't see any instruction file in changed files. 
I expect entry in main `readme.md` about testing.

**Comment by adrianbroher** [comment]
The README.md is not the right place for this kind of documentation. It should be an entry point for every target audience so in it's current state it is already too large. No need to cramp in more stuff for a very specialized part of the target audience (developers/contributors).

**Comment by o01eg** [comment]
I suppose `CONTRIBUTING.md` is the right place for testing description. But project testing is immature now.

**Comment by genestack-solomatin** [comment]
Any documentation in wrong place is better than no documentation. 

If I understand correctly how this work, this is useful for me, but I dont know how to use it.

> I suppose CONTRIBUTING.md is the right place for testing description.

May be just create testing.md in root folder and leave other documentation improvement out of scope of this PR.

> But project testing is immature now.

First steps are done.

**Comment by adrianbroher** [comment]
Using `CONTRIBUTING.md` is fine. Just add a new section related to testing.

**Comment by o01eg** [comment]
@adrianbroher I'm not sure what does need to be written about testing. There are also other tests except mine.

**Comment by adrianbroher** [comment]
@o01eg I'm the wrong person to anwer that question. Ask @genestack-solomatin

**Comment by Cjkjvfnby** [comment]
> I'm not sure what does need to be written about testing.

- what is tested 
- how it should be run
- arguments, platform specific

**Comment by o01eg** [comment]
Rebased.

**Comment by o01eg** [comment]
What if merge it into 0.4.8 branch too so release will be tested too?

**Comment by geoffthemedio** [comment]
no. adding new / potentially problematic stuff immediately before release not appropriate.

**Comment by Vezzra** [comment]
Considering the amount of changes to the codebase, this isn't a purely "infrastructure" PR anyway.

I'm going to add "component:internal" and a proper milestone (post release).

**Pull request merged**